src/aspire/source/__init__.py

```diff
@@ -1,57 +1,98 @@
+    ImageSource. The number of rows in this metadata table will equal the total no. of images supported by this
+    ImageSource (available as the 'n' attribute), though reading/writing of images is usually done in chunks.
+
+    This metadata table is implemented as a pandas DataFrame.
+
+    The 'values' in this metadata table are usually primitive types (floats/ints/strings) that are suitable
+    for being read from STAR files, and being written to STAR files. The columns corresponding to these fields
+    begin with a single underscore '_'.
+
+    In addition, the metadata table may also contain references to Python objects.
+    'Filter' objects, for example, are stored in this metadata table as references  to unique Filter objects that
+    correspond to images in this ImageSource. Several rows of metadata may end up containing a reference to a small
+    handful of unique Filter objects, depending on the values found in other columns (identical Filter
+    objects, depending on unique CTF values found for _rlnDefocusU/_rlnDefocusV etc.
+    """
+
+    """
+    The metadata_fields dictionary below specifies default data types of certain key fields used in the codebase.
+    The STAR file used to initialize subclasses of ImageSource may well contain other columns not found below; these
+    additional columns are available when read, and they default to the pandas data type 'object'.
+    """
+    metadata_fields = {
+        '_rlnVoltage': float,
+        '_rlnDefocusU': float,
+        '_rlnDefocusV': float,
+        '_rlnDefocusAngle': float,
+        '_rlnSphericalAberration': float,
+        '_rlnDetectorPixelSize': float,
+        '_rlnCtfFigureOfMerit': float,
+        '_rlnMagnification': float,
+        '_rlnAmplitudeContrast': float,
+        '_rlnImageName': str,
+        '_rlnOriginalName': str,
+        '_rlnCtfImage': str,
+        '_rlnCoordinateX': float,
+        '_rlnCoordinateY': float,
+        '_rlnCoordinateZ': float,
+        '_rlnNormCorrection': float,
+        '_rlnMicrographName': str,
+        '_rlnGroupName': str,
+        '_rlnGroupNumber': str,
+        '_rlnOriginX': float,
+        '_rlnOriginY': float,
+        '_rlnAngleRot': float,
+        '_rlnAngleTilt': float,
+        '_rlnAnglePsi': float,
+        '_rlnClassNumber': int,
+        '_rlnLogLikeliContribution': float,
+        '_rlnRandomSubset': int,
+        '_rlnParticleName': str,
+        '_rlnOriginalParticleName': str,
+        '_rlnNrOfSignificantSamples': float,
+        '_rlnNrOfFrames': int,
+        '_rlnMaxValueProbDistribution': float
     }
 
-    # All metadata fields are strings by default, specify any overrides here.
-    # Use the renamed metadata field name (i.e the 'values' in _metadata_aliases) to specify these.
-    _metadata_types = {}
-    # ----------------------------------------------------
+    @classmethod
+    def from_image(cls, im):
```

Comment:
Why not make this a subclass, say `ArrayImageSource`?

Reply:
A new class `ArrayImageSource` added.